### PR TITLE
NEXUS-23353: Fixed download and upload urls file generation

### PR DESCRIPTION
**Pull request by igorudovika**
Fixed download and upload urls file generation

### CODE REVIEW

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/internal/hosted/v1/ConanHostedFacet.java [hunk 1]
```diff
@@ -138,12 +141,28 @@ protected void doPutArchive(final ConanCoords coord,
...
+    String repositoryUrl = getRepository().getUrl();
+    Map<String, String> downloadUrls = assetsToUpload.stream()
+        .map(fileName -> Arrays.stream(AssetKind.values())
+            .filter(assetKind -> assetKind.getFilename().equals(fileName))
+            .findAny()
+            .orElseThrow(
+                () -> new IllegalArgumentException("Unknown asset kind for uploading file: " + fileName))
+        )
+        .collect(Collectors.toMap(AssetKind::getFilename,
+            x -> repositoryUrl + "/" + ConanHostedHelper.getHostedAssetPath(coords, x)));
```
Comment (mattjohnson) [inline nexus-repository-conan/src/main/java/org/sonatype/repository/conan/internal/hosted/v1/ConanHostedFacet.java:164]: Minor: I'd find a for loop more readable over nested streams.
```
for (String fileName : assetsToUpload) {
  AssetKind assetKind = AssetKind.valueFromFileName(fileName);
  if (assetKind == null) {
    throw new IllegalArgumentException("Unknown asset kind for uploading file: " + fileName);
  }
  downloadUrls.put(fileName, repositoryUrl + "/" + ConanHostedHelper.getHostedAssetPath(coords, assetKind));
}
```
Adding a method to `AssetKind` to do the conversation between filename and `AssetKind`.
```
static AssetKind valueFromFileName(final String filename) {
  values().find { it.filename == filename }
}
```

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/internal/hosted/v1/HostedHandlers.java [hunk 2]
```diff
@@ -48,14 +55,30 @@
   public static final Handler uploadUrl = context -> {
     State state = context.getAttributes().require(State.class);
     ConanCoords coord = ConanHostedHelper.convertFromState(state);
-    String downloadUrlAsJson = context.getRepository()
-        .facet(ConanHostedFacet.class)
-        .getDownloadUrlAsJson(coord);
+    ObjectMapper objectMapper = new ObjectMapper();
```
Comment (mattjohnson) [inline nexus-repository-conan/src/main/java/org/sonatype/repository/conan/internal/hosted/v1/HostedHandlers.java:58]: Minor: `ObjectMapper` can be a heavier class. Consider adding a private static field `private static ObjectReader objectReader = new ObjectMapper().reader();` and use the reader directly, which is thread safe a lighter weight. Might not be a big deal if this handler is not expected to be used frequently.

Also using the reader I don't think you'll need the intermediate `TypeReference` to read the value.

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/internal/hosted/ConanHostedMetadataFacetSupport.java [hunk 3]
```diff
@@ -60,32 +61,30 @@
...
+      final ConanCoords coords)
   {
+    Repository repository = getRepository();
+    StorageTx tx = UnitOfWork.currentTx();
     return assetKinds
-        .stream()
+        .stream().filter(x -> tx.assetExists(ConanHostedHelper.getHostedAssetPath(coords, x), repository))
         .collect(toMap(AssetKind::getFilename,
-            x -> repositoryUrl + "/" + ConanHostedHelper.getHostedAssetPath(coords, x)));
+            x -> repository.getUrl() + "/" + ConanHostedHelper.getHostedAssetPath(coords, x)));
```
[the remark this answers is not in the document]
Comment (igorudovika) [reply nexus-repository-conan/src/main/java/org/sonatype/repository/conan/internal/hosted/ConanHostedMetadataFacetSupport.java]: Fixed 25edb0efc472f8b9787965230222b039736cda15

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/internal/hosted/ConanHostedMetadataFacetSupport.java [hunk 4]
```diff
@@ -60,32 +61,30 @@
...
+  private Map<String, String> generateDownloadUrls(
       final List<AssetKind> assetKinds,
-      final ConanCoords coords, final String repositoryUrl)
+      final ConanCoords coords)
   {
+    Repository repository = getRepository();
+    StorageTx tx = UnitOfWork.currentTx();
     return assetKinds
-        .stream()
+        .stream().filter(x -> tx.assetExists(ConanHostedHelper.getHostedAssetPath(coords, x), repository))
```
[the remark this answers is not in the document]
Comment (igorudovika) [reply nexus-repository-conan/src/main/java/org/sonatype/repository/conan/internal/hosted/ConanHostedMetadataFacetSupport.java]: Fixed 25edb0efc472f8b9787965230222b039736cda15

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/internal/hosted/ConanHostedMetadataFacetSupport.java [hunk 5]
```diff
@@ -60,32 +61,30 @@
...
 
-  public String generateDownloadPackagesUrlsAsJson(
-      final ConanCoords coords,
-      final String repositoryUrl) throws JsonProcessingException
+  public String generateDownloadPackagesUrlsAsJson(final ConanCoords coords)
+      throws JsonProcessingException
   {
     Map<String, String> downloadUrls = generateDownloadUrls(DOWNLOAD_URL_PACKAGE_ASSET_KINDS,
-        coords, repositoryUrl);
+        coords);
```
[the remark this answers is not in the document]
Comment (igorudovika) [reply nexus-repository-conan/src/main/java/org/sonatype/repository/conan/internal/hosted/ConanHostedMetadataFacetSupport.java]: Fixed 25edb0efc472f8b9787965230222b039736cda15

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/internal/hosted/v1/HostedHandlers.java (hunk 2 above)
Comment (igorudovika) [reply nexus-repository-conan/src/main/java/org/sonatype/repository/conan/internal/hosted/v1/HostedHandlers.java]: Thanks, fixed 25edb0efc472f8b9787965230222b039736cda15

File: nexus-repository-conan/src/main/java/org/sonatype/repository/conan/internal/hosted/v1/ConanHostedFacet.java (hunk 1 above)
Comment (igorudovika) [reply nexus-repository-conan/src/main/java/org/sonatype/repository/conan/internal/hosted/v1/ConanHostedFacet.java]: Ok, let use such approach 9007e590aa0043bb8f7f78080657c842661f3c94

**Pull request merged**